File: frontend/lib/soap-note-generator.ts
/**
 * Template Method pattern for clinical SOAP note generation (NFR-03 extensibility).
 *
 * SoapNoteGenerator is the abstract class that defines the fixed algorithm:
 *   1. Build a specialty-aware prompt  (invariant — done by base class)
 *   2. Call Claude Sonnet              (invariant — done by base class)
 *   3. Parse and normalise fields      (invariant — done by base class)
 *
 * Subclasses override only `templateName`, `fields`, and the optional
 * `specialtyContext()` hook.  Adding a new specialty requires only a new
 * subclass and one line in NoteGeneratorFactory — the API route is untouched.
 */

import Anthropic from "@anthropic-ai/sdk"

// ── Shared helper ─────────────────────────────────────────────────────────────

function extractJson(text: string): Record<string, string> {
  const match = text.match(/\{[\s\S]*\}/)
  if (!match) return {}
  try { return JSON.parse(match[0]) as Record<string, string> } catch { return {} }
}

// ── Abstract base ─────────────────────────────────────────────────────────────

export abstract class SoapNoteGenerator {
  abstract readonly templateName: string
  abstract readonly fields: readonly string[]

  /**
   * Template method — fixed skeleton; subclasses customise via overrides.
   * Returns a record with every field key present (empty string if not found).
   */
  async generate(
    transcript: string,
    client: Anthropic,
  ): Promise<Record<string, string>> {
    const raw = await this.callModel(transcript, client)
    return this.normaliseFields(raw)
  }

  /** Hook — subclasses may override to inject specialty context into the prompt. */
  protected specialtyContext(): string {
    return ""
  }

  private async callModel(
    transcript: string,
    client: Anthropic,
  ): Promise<Record<string, string>> {
    const ctx = this.specialtyContext()
    const res = await client.messages.create({
      model:      "claude-sonnet-4-6",
      max_tokens: 2048,
      messages: [{
        role: "user",
        content: [
          `You are a medical scribe${ctx ? ` specialising in ${ctx}` : ""}.`,
          `Generate a structured clinical note from this consultation transcript.`,
          ``,
          `TRANSCRIPT:`,
          transcript,
          ``,
          `Return a JSON object with EXACTLY these keys: ${this.fields.join(", ")}`,
          `Fill each key with concise, clinical prose from the transcript.`,
          `Leave a key as "" if the content is not mentioned.`,
          `Respond with JSON only — no markdown, no explanation.`,
        ].join("\n"),
      }],
    })

    const text = res.content[0].type === "text" ? res.content[0].text : "{}"
    return extractJson(text)
  }

  private normaliseFields(raw: Record<string, string>): Record<string, string> {
    const result: Record<string, string> = {}
    for (const f of this.fields) result[f] = raw[f] ?? ""
    return result
  }
}

// ── Concrete generators ───────────────────────────────────────────────────────

export class GeneralOpdNoteGenerator extends SoapNoteGenerator {
  readonly templateName = "general_opd"
  readonly fields = ["subjective", "objective", "assessment", "diagnosis", "prescription", "advice", "follow_up"] as const
}

export class MentalHealthNoteGenerator extends SoapNoteGenerator {
  readonly templateName = "mental_health_soap"
  readonly fields = ["subjective", "objective", "assessment", "plan", "safety_assessment"] as const
  protected specialtyContext() { return "mental health and psychiatry" }
}

export class PhysiotherapyNoteGenerator extends SoapNoteGenerator {
  readonly templateName = "physiotherapy"
  readonly fields = ["subjective", "objective", "assessment", "treatment", "home_exercise_program", "plan"] as const
  protected specialtyContext() { return "physiotherapy and rehabilitation" }
}

export class PediatricNoteGenerator extends SoapNoteGenerator {
  readonly templateName = "pediatric"
  readonly fields = ["subjective", "objective", "assessment", "plan", "parent_instructions", "follow_up"] as const
  protected specialtyContext() { return "paediatrics" }
}

export class CardiologyNoteGenerator extends SoapNoteGenerator {
  readonly templateName = "cardiology"
  readonly fields = ["subjective", "objective", "assessment", "plan", "medications", "follow_up"] as const
  protected specialtyContext() { return "cardiology and cardiovascular medicine" }
}

export class SurgicalFollowupNoteGenerator extends SoapNoteGenerator {
  readonly templateName = "surgical_followup"
  readonly fields = ["wound_assessment", "subjective", "objective", "assessment", "plan", "next_review"] as const
  protected specialtyContext() { return "post-operative surgical follow-up" }
}

// ── Factory ───────────────────────────────────────────────────────────────────

const GENERATORS: SoapNoteGenerator[] = [
  new GeneralOpdNoteGenerator(),
  new MentalHealthNoteGenerator(),
  new PhysiotherapyNoteGenerator(),
  new PediatricNoteGenerator(),
  new CardiologyNoteGenerator(),
  new SurgicalFollowupNoteGenerator(),
]

export class NoteGeneratorFactory {
  /** Return the generator for the given template name, defaulting to General OPD. */
  static get(templateName: string): SoapNoteGenerator {
    return GENERATORS.find(g => g.templateName === templateName)
      ?? new GeneralOpdNoteGenerator()
  }

  /** All registered template names — used for template-detection prompts. */
  static templateNames(): string[] {
    return GENERATORS.map(g => g.templateName)
  }
}

File: backend/java/src/main/java/com/scribehealth/pattern/template/SoapNoteGenerator.java
package com.scribehealth.pattern.template;

import java.util.HashMap;
import java.util.Map;

/**
 * Template Method pattern for clinical SOAP note generation (NFR-03 extensibility).
 *
 * SoapNoteGenerator is the abstract class that defines the fixed algorithm:
 *   1. Build a specialty-aware prompt  (invariant — done by base class)
 *   2. Call AI Model                  (invariant — done by base class)
 *   3. Parse and normalize fields     (invariant — done by base class)
 *
 * Subclasses override only `templateName`, `fields`, and the optional
 * `specialtyContext()` hook. Adding a new specialty requires only a new
 * subclass and one line in NoteGeneratorFactory — the API route is untouched.
 */
public abstract class SoapNoteGenerator {
    
    /**
     * Returns the template name for this generator.
     * @return template identifier (e.g., "general_opd", "cardiology")
     */
    public abstract String getTemplateName();
    
    /**
     * Returns the list of field names for this template.
     * @return array of field names that will be present in the generated note
     */
    public abstract String[] getFields();
    
    /**
     * Template method — fixed skeleton; subclasses customize via overrides.
     * Returns a map with every field key present (empty string if not found).
     *
     * @param transcript the consultation transcript to generate from
     * @param aiClient the AI client for calling the model
     * @return map of field names to generated content
     */
    public final Map<String, String> generate(String transcript, AiClient aiClient) {
        String raw = callModel(transcript, aiClient);
        return normalizeFields(raw);
    }
    
    /**
     * Hook — subclasses may override to inject specialty context into the prompt.
     * @return specialty context string (e.g., "cardiology and cardiovascular medicine")
     */
    protected String specialtyContext() {
        return "";
    }
    
    /**
     * Calls the AI model to generate the note content.
     * This is a private method that implements the invariant algorithm step.
     */
    private String callModel(String transcript, AiClient aiClient) {
        String ctx = specialtyContext();
        StringBuilder prompt = new StringBuilder();
        
        prompt.append("You are a medical scribe");
        if (!ctx.isEmpty()) {
            prompt.append(" specialising in ").append(ctx);
        }
        prompt.append(".\n\n");
        prompt.append("Generate a structured clinical note from this consultation transcript.\n\n");
        prompt.append("TRANSCRIPT:\n");
        prompt.append(transcript);
        prompt.append("\n\n");
        prompt.append("Return a JSON object with EXACTLY these keys: ");
        prompt.append(String.join(", ", getFields()));
        prompt.append("\nFill each key with concise, clinical prose from the transcript.\n");
        prompt.append("Leave a key as \"\" if the content is not mentioned.\n");
        prompt.append("Respond with JSON only — no markdown, no explanation.");
        
        return aiClient.generateText(prompt.toString());
    }
    
    /**
     * Normalizes the fields to ensure all expected fields are present.
     * This is a private method that implements the invariant algorithm step.
     */
    private Map<String, String> normalizeFields(String raw) {
        Map<String, String> result = new HashMap<>();
        Map<String, String> parsed = extractJson(raw);
        
        for (String field : getFields()) {
            result.put(field, parsed.getOrDefault(field, ""));
        }
        
        return result;
    }
    
    /**
     * Extracts JSON from a text string.
     * Looks for the first JSON object in the text.
     */
    private Map<String, String> extractJson(String text) {
        if (text == null || text.isEmpty()) {
            return new HashMap<>();
        }
        
        // Find the first { and last }
        int start = text.indexOf('{');
        int end = text.lastIndexOf('}');
        
        if (start == -1 || end == -1 || start >= end) {
            return new HashMap<>();
        }
        
        String json = text.substring(start, end + 1);
        
        // Simple JSON parsing for flat objects with string values
        Map<String, String> result = new HashMap<>();
        
        // Remove outer braces and split by commas (naive approach for simple JSON)
        String content = json.substring(1, json.length() - 1).trim();
        if (content.isEmpty()) {
            return result;
        }
        
        // Split by comma, but be careful about commas inside quotes
        String[] pairs = content.split(",(?=([^\"]*\"[^\"]*\")*[^\"]*$)");
        
        for (String pair : pairs) {
            String[] kv = pair.split(":", 2);
            if (kv.length == 2) {
                String key = kv[0].trim().replaceAll("^\"|\"$", "");
                String value = kv[1].trim().replaceAll("^\"|\"$", "");
                result.put(key, value);
            }
        }
        
        return result;
    }
}
